File: src/ch04/code_6.rs
fn exist_if_sum_pair(i: usize, target: i32, nums: &Vec<i32>) -> bool {
    if i == 0 {
        return target == 0;
    }
    if exist_if_sum_pair(i-1, target, nums) {
        return true;
    }
    if exist_if_sum_pair(i-1, target - nums[i-1], nums) {
        return true;
    }
    false
}


fn main() {
    let target: i32 = 10;
    let nums: Vec<i32> = vec![1, 2, 4, 5, 11];
    let result: bool = exist_if_sum_pair(nums.len(), target, &nums);

    println!("{}", result);
}

File: src/ch04/code_6.py
from typing import List

from rich import print


def exist_if_sum_pair(i: int, target: int, nums: List[int]) -> bool:
    if i == 0:
        return target == 0
    if exist_if_sum_pair(i=i-1, target=target, nums=nums):
        return True
    if exist_if_sum_pair(i=i-1, target=target-nums[i-1], nums=nums):
        return True
    return False


if __name__ == '__main__':
    target: int = 10
    nums: List[int] = [1, 2, 4, 5, 11]
    result: bool = exist_if_sum_pair(i=len(nums), target=target, nums=nums)

    print(result)
